### PR TITLE
style(js): Enable `plugin:react/recommended` eslint rules

### PR DESCRIPTION
**Pull request by mcous**
## overview

As discussed in #1788, this PR enables the `plugin:react/recommended` rule and resolves (or, in the case of one error, ignores) existing lint errors.

## changelog

- style(js): Enable `plugin:react/recommended` eslint rules 

## review requests

This change should **have no affect** on app, component library, and protocol designer. Please check the touched files to make sure nothing broke.

I will annotate each change with the rule for clarity

### CODE REVIEW

File: app/src/components/ConfirmTipProbeModal/Contents.js [hunk 1]
```diff
@@ -5,7 +5,7 @@ import * as React from 'react'
 import removeTrashSrc from './img/trash@3x.png'
 import styles from './Contents.css'
 
-export default function () {
+export default function Contents () {
```
Comment (mcous) [inline app/src/components/ConfirmTipProbeModal/Contents.js]: [react/display-name](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/display-name.md)

File: app/src/components/RunLog/CommandList.js [hunk 2]
```diff
@@ -9,7 +9,8 @@ import styles from './styles.css'
 
 export default class CommandList extends Component {
   componentDidUpdate () {
-    if (this.refs.ensureVisible) this.refs.ensureVisible.scrollIntoView(true)
+    // TODO(mc, 2018-07-24): use new refs
+    if (this.refs.ensureVisible) this.refs.ensureVisible.scrollIntoView(true) // eslint-disable-line react/no-string-refs
```
Comment (mcous) [inline app/src/components/RunLog/CommandList.js]: [react/no-string-refs](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-string-refs.md)

File: app/src/components/RunPanel/RunControls.js [hunk 3]
```diff
@@ -16,7 +16,7 @@ type RunProps = {
   onResumeClick: () => mixed,
   onResetClick: () => mixed,
 }
-export default function (props: RunProps) {
+export default function RunControls (props: RunProps) {
```
Comment (mcous) [inline app/src/components/RunPanel/RunControls.js]: [react/display-name](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/display-name.md)

File: app/src/pages/Robots/index.js [hunk 4]
```diff
@@ -53,13 +53,12 @@ function Robots (props: Props) {
...
         path={`${path}/instruments`}
-        render={(props) => (
-          <InstrumentSettings {...props} robot={robot} />
-        )} />
+        render={props => (<InstrumentSettings {...props} robot={robot} />)}
+      />
+      <Route
+        path={path}
+        render={() => (<RobotSettings robot={robot} />)}
       />
```
Comment (mcous) [inline app/src/pages/Robots/index.js]: [react/no-unescaped-entities](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-unescaped-entities.md)

Happy this typo was caught 🎉

File: protocol-designer/src/components/IngredientPropertiesForm.js [hunk 5]
```diff
@@ -38,7 +38,7 @@ type FieldProps = {|
 |}
 
 const makeInputField = (args: {setSubstate: SetSubstate, getSubstate: GetSubstate}) =>
-  (props: FieldProps) => { /* otherProps */
+  function InputFieldFactory (props: FieldProps) { /* otherProps */
```
Comment (mcous) [inline protocol-designer/src/components/IngredientPropertiesForm.js]: [react/display-name](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/display-name.md)

File: protocol-designer/src/components/IngredientPropertiesForm.js [hunk 6]
```diff
@@ -173,7 +173,8 @@ class IngredientPropertiesForm extends React.Component<Props, State> {
     }
   }
 
-  componentWillReceiveProps (nextProps: Props) {
+  // TODO(mc, 2018-07-24): use a different lifecycle hook
+  UNSAFE_componentWillReceiveProps (nextProps: Props) {
```
Comment (mcous) [inline protocol-designer/src/components/IngredientPropertiesForm.js]: [react/no-deprecated](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-deprecated.md)

File: protocol-designer/src/components/labware/ClickableText.js [hunk 7]
```diff
@@ -10,7 +10,7 @@ type Props = {
   onClick?: (e: SyntheticEvent<*>) => void,
 }
 
-export default function (props: Props) {
+export default function ClickableText (props: Props) {
```
Comment (mcous) [inline protocol-designer/src/components/labware/ClickableText.js]: [react/display-name](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/display-name.md)

File: protocol-designer/src/components/modals/MoreOptionsModal.js [hunk 8]
```diff
@@ -20,7 +20,7 @@ type Props = {
   hideModal: boolean
 }
 
-export default function (props: Props) {
+export default function MoreOptionsModal (props: Props) {
```
Comment (mcous) [inline protocol-designer/src/components/modals/MoreOptionsModal.js]: [react/display-name](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/display-name.md)

File: protocol-designer/src/components/modals/NewFileModal/index.js [hunk 9]
```diff
@@ -65,7 +65,8 @@ export default class NewFileModal extends React.Component<Props, State> {
     this.state = initialState
   }
 
-  componentWillReceiveProps (nextProps: Props) {
+  // TODO(mc, 2018-07-24): use a different lifecycle hook
+  UNSAFE_componentWillReceiveProps (nextProps: Props) {
```
Comment (mcous) [inline protocol-designer/src/components/modals/NewFileModal/index.js]: [react/no-deprecated](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-deprecated.md)

File: protocol-designer/src/components/modals/NewFileModal/index.js [hunk 10]
```diff
@@ -163,12 +164,12 @@ const BetaRestrictions = () => (
     <h3>Beta Pipette Restrictions:</h3>
     <ol>
       <li>
-        You can't change your pipette selection later. If in doubt go for
+        You can&apos;t change your pipette selection later. If in doubt go for
```
Comment (mcous) [inline protocol-designer/src/components/modals/NewFileModal/index.js]: [react/no-unescaped-entities](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-unescaped-entities.md)

These can also be fixed via braces interpolation (e.g. `{"'"}`); not sure what everyone's preference is

File: protocol-designer/src/components/modals/NewFileModal/index.js [hunk 11]
```diff
@@ -163,12 +164,12 @@ const BetaRestrictions = () => (
...
         smaller pipettes. The Protocol Designer automatically breaks up transfer
         volumes that exceed pipette capacity into multiple transfers.
       </li>
       <li>
-        Pipettes can't share tip racks. There needs to be at least 1 tip rack per
+        Pipettes can&apos;t share tip racks. There needs to be at least 1 tip rack per
```
Comment (mcous) [inline protocol-designer/src/components/modals/NewFileModal/index.js]: [react/no-unescaped-entities](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-unescaped-entities.md)

File: protocol-designer/src/containers/LabwareNameEditForm.js [hunk 12]
```diff
@@ -41,7 +41,8 @@ class LabwareNameEditForm extends React.Component<Props, State> {
     this.setState({name: e.target.value})
   }
 
-  componentWillReceiveProps (nextProps: Props) {
+  // TODO(mc, 2018-07-24): use a different lifecycle hook
+  UNSAFE_componentWillReceiveProps (nextProps: Props) {
```
Comment (mcous) [inline protocol-designer/src/containers/LabwareNameEditForm.js]: [react/no-deprecated](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-deprecated.md)

File: protocol-designer/src/components/modals/NewFileModal/index.js (hunk 10 above)
Comment (Kadee80) [reply protocol-designer/src/components/modals/NewFileModal/index.js]: Personally prefer the html entity option.

File: protocol-designer/src/components/modals/NewFileModal/index.js (hunk 9 above)
Comment (IanLondon) [reply protocol-designer/src/components/modals/NewFileModal/index.js]: This seems to break behavior of NewFileModal - opening it a second time doesn't clear it out. Could we keep the TODO and add a lint ignore for this?

File: protocol-designer/src/components/IngredientPropertiesForm.js (hunk 6 above)
Comment (IanLondon) [reply protocol-designer/src/components/IngredientPropertiesForm.js]: breaks - changing well selection now fails to update form fields

File: protocol-designer/src/containers/LabwareNameEditForm.js (hunk 12 above)
Comment (IanLondon) [reply protocol-designer/src/containers/LabwareNameEditForm.js]: breaks: opening the form doesn't populate

**Comment by mcous** [comment]
@IanLondon fixed the `UNSAFE_...` thing.

W.R.T. comma dangle: `only-multiline` allows comma dangle without enforcing them (which is `always-multiline`). Since it's an `eslint --fix`able rule, I'd be more in favor of ripping the bandaid off in one go, but if people would prefer to do it gradually that's cool, too. The only downside to gradual I can think of is the potential for unrelated style changes in various PRs as we switch over

File: .eslintrc.js [hunk 13]
```diff
@@ -24,5 +27,11 @@ module.exports = {
     node: true,
     browser: true,
     jest: true
+  },
+
+  settings: {
+    react: {
+      version: '16.2'
```
Comment (IanLondon) [inline .eslintrc.js]: Is there a sensible way to read this from the `package.json` so we don't lose sync? Not a big deal

Comment (mcous) [reply .eslintrc.js]: I couldn't think of one, given how we keep a react in a devDep at root, peerDep in components, and dep in app / protocol-designer. Let's figure something out when we update React?

**Pull request merged**